bsms.mediawiki

```diff
@@ -0,0 +1,176 @@
+
+=====Signer=====
+
+* The Signer initiates a new secure multisig setup session by setting the <tt>TOKEN</tt>. The Signer derives an <tt>ENCRYPTION_KEY</tt> from the <tt>TOKEN</tt>. The Signer can keep the session open until a different value for the <tt>TOKEN</tt> is set.
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Optionally, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line is a text description of the key, 80 characters maximum. The fourth line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first three lines. The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted.
+* Finally, the Signer encrypts the key record with <tt>ENCRYPTION_KEY</tt>. The Signer calculates the 256-bit Initialization Vector (<tt>IV</tt>) for the XPUB, and appends it to the end of the encrypted text.
+
+====Round 2====
+
+=====Coordinator=====
+
+* The Coordinator gathers key records from all participating Signers. Abort the setup if there are fewer than <tt>N</tt> unique key records when the wallet setup session expires.
+* For each key record, the Coordinator extracts the <tt>IV</tt> from the data, then decrypts the ciphertext using <tt>ENCRYPTION_KEY</tt> and <tt>IV</tt>.
+* The Coordinator verifies that the included <tt>IV</tt> is valid given the decrypted data.
+* The Coordinator verifies that the included <tt>SIG</tt> is valid given the <tt>KEY</tt>.
+* If all key records look good, the Coordinator fills in all necessary information to generate a descriptor record, which is simply the descriptor string plus a <tt>CHECKSUM</tt>, all in one line. The <tt>CHECKSUM</tt> has [https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#checksums BECH32 encoding].
+* The Coordinator encrypts the descriptor record with <tt>ENCRYPTION_KEY</tt>. The Coordinator calculates the 256-bit <tt>IV</tt> for the descriptor, and appends it to the end of the encrypted text.
+* The Coordinator sends the encrypted descriptor record to all participating Signers.
+
+=====Signer=====
+
+* The Signer imports the descriptor record.
+* The Signer extracts the <tt>IV</tt> from the data, then decrypts the ciphertext using <tt>ENCRYPTION_KEY</tt> (derived from the open session) and <tt>IV</tt>.
+* The Signer verifies that the included <tt>IV</tt> is valid given the decrypted data.
```

Comment:
Reinterprets the IV as HMAC output and validates, 'IV is valid' is a confusing phrase, IV is just a nonce